beef_bump.go

```diff
@@ -0,0 +1,144 @@
+package bux
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/libsv/go-bt/v2"
+)
+
+func calculateMergedBUMP(txs []*Transaction) (BUMPs, error) {
+	bumps := make(map[uint64][]BUMP)
+	mergedBUMPs := make(BUMPs, 0)
+
+	for _, tx := range txs {
+		if tx.BUMP.BlockHeight == 0 || len(tx.BUMP.Path) == 0 {
+			continue
+		}
+
+		bumps[tx.BlockHeight] = append(bumps[tx.BlockHeight], tx.BUMP)
+	}
+
+	// ensure that BUMPs are sorted by block height and will always be put in beef in the same order
+	mapKeys := make([]uint64, 0, len(bumps))
+	for k := range bumps {
+		mapKeys = append(mapKeys, k)
+	}
+	sort.Slice(mapKeys, func(i, j int) bool { return mapKeys[i] < mapKeys[j] })
+
+	for _, k := range mapKeys {
+		bump, err := CalculateMergedBUMP(bumps[k])
+		if err != nil {
+			return nil, fmt.Errorf("Error while calculating Merged BUMP: %s", err.Error())
+		}
+		if bump == nil {
+			continue
+		}
+		mergedBUMPs = append(mergedBUMPs, bump)
+	}
+
+	return mergedBUMPs, nil
+}
+
+func validateBumps(bumps BUMPs) error {
+	if len(bumps) == 0 {
+		return errors.New("empty bump paths slice")
+	}
+
+	for _, p := range bumps {
+		if len(p.Path) == 0 {
+			return errors.New("one of bump path is empty")
+		}
+	}
+
+	return nil
+}
+
+func prepareBUMPFactors(ctx context.Context, tx *Transaction, store TransactionGetter) ([]*bt.Tx, []*Transaction, error) {
+	btTxsNeededForBUMP, txsNeededForBUMP, err := initializeRequiredTxsCollection(tx)
```

Comment:
I have two collections here because I need *Transaction and *bt.Tx during the whole process. So to avoid multiple conversions I do it once and then pass.

It will be removed when we move it to go.bt/go.bc and one structure will have all required fields like BUMP, Inputs etc.